File: src/frontend/src/services/contest.ts
/**
 * 竞赛相关 API 服务
 */

import request from '@/utils/request'
import type { Contest, ContestListItem, ContestProblem, RankItem } from '@/types'

interface GetContestsParams {
  page?: number
  page_size?: number
  status?: string
  keyword?: string
}

interface GetContestsResponse {
  items: ContestListItem[]
  total: number
  page: number
  page_size: number
}

interface CreateContestData {
  title: string
  description?: string
  start_time: string
  end_time: string
  duration: number
  contest_type?: 'public' | 'private'
  rule_type?: 'acm' | 'oi'
  password?: string
  is_visible?: boolean
  real_time_rank?: boolean
  freeze_time?: number
}

interface UpdateContestData {
  title?: string
  description?: string
  start_time?: string
  end_time?: string
  duration?: number
  contest_type?: 'public' | 'private'
  rule_type?: 'acm' | 'oi'
  password?: string
  is_visible?: boolean
  real_time_rank?: boolean
  freeze_time?: number
}

interface AddContestProblemData {
  problem_id: number
  display_id: string
  score?: number
}

interface RegisterContestData {
  password?: string
}

interface GetRankResponse {
  items: RankItem[]
  total: number
  is_frozen: boolean
}

export const contestApi = {
  /**
   * 获取竞赛列表
   */
  getContests: async (params: GetContestsParams = {}): Promise<GetContestsResponse> => {
    return request.get('/contests', { params })
  },

  /**
   * 创建竞赛
   */
  createContest: async (data: CreateContestData): Promise<Contest> => {
    return request.post('/contests', data)
  },

  /**
   * 获取竞赛详情
   */
  getContest: async (id: number): Promise<Contest> => {
    return request.get(`/contests/${id}`)
  },

  /**
   * 更新竞赛
   */
  updateContest: async (id: number, data: UpdateContestData): Promise<Contest> => {
    return request.put(`/contests/${id}`, data)
  },

  /**
   * 删除竞赛
   */
  deleteContest: async (id: number): Promise<void> => {
    return request.delete(`/contests/${id}`)
  },

  /**
   * 获取竞赛题目列表
   */
  getContestProblems: async (contestId: number): Promise<ContestProblem[]> => {
    return request.get(`/contests/${contestId}/problems`)
  },

  /**
   * 添加竞赛题目
   */
  addContestProblem: async (contestId: number, data: AddContestProblemData): Promise<ContestProblem> => {
    return request.post(`/contests/${contestId}/problems`, data)
  },

  /**
   * 移除竞赛题目
   */
  removeContestProblem: async (contestId: number, problemId: number): Promise<void> => {
    return request.delete(`/contests/${contestId}/problems/${problemId}`)
  },

  /**
   * 报名参赛
   */
  registerContest: async (contestId: number, data: RegisterContestData = {}): Promise<{ message: string }> => {
    return request.post(`/contests/${contestId}/register`, data)
  },

  /**
   * 获取竞赛排行榜
   */
  getContestRank: async (contestId: number, page: number = 1, pageSize: number = 50): Promise<GetRankResponse> => {
    return request.get(`/contests/${contestId}/rank`, {
      params: { page, page_size: pageSize }
    })
  },
}

export default contestApi

File: src/backend/app/schemas/contest.py
"""
竞赛相关的 Pydantic schemas
"""

from datetime import datetime
from typing import Optional
from pydantic import BaseModel, Field, ConfigDict

from app.models.contest import ContestType, ContestStatus, RuleType


# ============ Contest Schemas ============

class ContestBase(BaseModel):
    """竞赛基础 Schema"""
    title: str = Field(..., min_length=1, max_length=200)
    description: Optional[str] = None
    start_time: datetime
    end_time: datetime
    duration: int = Field(..., gt=0, description="持续时间（分钟）")
    contest_type: ContestType = ContestType.PUBLIC
    rule_type: RuleType = RuleType.ACM
    password: Optional[str] = Field(None, max_length=100)
    is_visible: bool = True
    real_time_rank: bool = True
    freeze_time: Optional[int] = Field(None, ge=0, description="封榜时间（分钟）")


class ContestCreate(ContestBase):
    """创建竞赛"""
    pass


class ContestUpdate(BaseModel):
    """更新竞赛"""
    title: Optional[str] = Field(None, min_length=1, max_length=200)
    description: Optional[str] = None
    start_time: Optional[datetime] = None
    end_time: Optional[datetime] = None
    duration: Optional[int] = Field(None, gt=0)
    contest_type: Optional[ContestType] = None
    rule_type: Optional[RuleType] = None
    password: Optional[str] = Field(None, max_length=100)
    is_visible: Optional[bool] = None
    real_time_rank: Optional[bool] = None
    freeze_time: Optional[int] = Field(None, ge=0)


class Contest(ContestBase):
    """竞赛详情"""
    model_config = ConfigDict(from_attributes=True)
    
    id: int
    created_by: int
    participant_count: int
    submission_count: int
    status: ContestStatus
    is_frozen: bool
    created_at: datetime
    updated_at: datetime


class ContestList(BaseModel):
    """竞赛列表项"""
    model_config = ConfigDict(from_attributes=True)
    
    id: int
    title: str
    start_time: datetime
    end_time: datetime
    duration: int
    contest_type: ContestType
    rule_type: RuleType
    participant_count: int
    status: ContestStatus
    created_at: datetime


class ContestListResponse(BaseModel):
    """竞赛列表响应"""
    items: list[ContestList]
    total: int
    page: int
    page_size: int


# ============ ContestProblem Schemas ============

class ContestProblemBase(BaseModel):
    """竞赛题目基础 Schema"""
    problem_id: int
    display_id: str = Field(..., max_length=10, description="题目序号（A, B, C...）")
    score: int = Field(100, ge=0, description="题目分数")


class ContestProblemCreate(ContestProblemBase):
    """添加竞赛题目"""
    pass


class ContestProblem(ContestProblemBase):
    """竞赛题目详情"""
    model_config = ConfigDict(from_attributes=True)
    
    id: int
    contest_id: int
    submission_count: int
    accepted_count: int
    created_at: datetime


class ContestProblemWithDetail(ContestProblem):
    """带题目详情的竞赛题目"""
    problem_title: str
    problem_difficulty: str


# ============ ContestParticipant Schemas ============

class ContestParticipantBase(BaseModel):
    """竞赛参与者基础 Schema"""
    pass


class ContestParticipantCreate(BaseModel):
    """报名参赛"""
    password: Optional[str] = None


class ContestParticipant(ContestParticipantBase):
    """竞赛参与者详情"""
    model_config = ConfigDict(from_attributes=True)
    
    id: int
    contest_id: int
    user_id: int
    is_registered: bool
    total_score: int
    solved_count: int
    total_time: int
    submission_count: int
    registered_at: datetime
    last_submission_at: Optional[datetime]


class RankItem(BaseModel):
    """排行榜项"""
    rank: int
    user_id: int
    username: str
    total_score: int = 0  # OI 规则
    solved_count: int = 0  # ACM 规则
    total_time: int = 0  # 总用时/罚时（分钟）
    submission_count: int = 0
    last_submission_at: Optional[datetime] = None


class RankListResponse(BaseModel):
    """排行榜响应"""
    items: list[RankItem]
    total: int
    is_frozen: bool
